src/Concordium/Client/GRPC2.hs

```diff
@@ -0,0 +1,2693 @@
+        return QueryTypes.NextUpdateSequenceNumbers{..}
+
+type IpAddress = Text
+type Peer = IpAddress
+type IpPort = Int16
+type IpSocketAddress = (IpAddress, IpPort)
+
+instance ToProto IpAddress where
+    type Output IpAddress = Proto.IpAddress
+    toProto ip = Proto.make $ ProtoFields.value .= ip
+
+instance ToProto IpPort where
+    type Output IpPort = Proto.Port
+    toProto ip = Proto.make $ ProtoFields.value .= fromIntegral ip
+
+instance FromProto Proto.IpAddress where
+    type Output' Proto.IpAddress = IpAddress
+    fromProto peer = peer ^? ProtoFields.value
+
+instance FromProto Proto.Port where
+    type Output' Proto.Port = IpPort
+    fromProto port = fromIntegral <$> port ^? ProtoFields.value
+
+instance FromProto Proto.BannedPeer where
+    type Output' Proto.BannedPeer = Peer
+    fromProto peer = fromProto =<< peer ^? ProtoFields.ipAddress
+
+instance FromProto Proto.BannedPeers where
+    type Output' Proto.BannedPeers = [Peer]
+    fromProto peers = mapM fromProto =<< peers ^? ProtoFields.peers
+
+instance FromProto Proto.InstanceStateValueAtKey where
+    type Output' Proto.InstanceStateValueAtKey = ByteString
+    fromProto bs = bs ^? ProtoFields.value
+
+instance FromProto Proto.Signature where
+    type Output' Proto.Signature = Signature
+    fromProto sig = Signature . BSS.toShort <$> (sig ^? ProtoFields.value)
+
+instance FromProto Proto.SignatureMap where
+    type Output' Proto.SignatureMap = Updates.UpdateInstructionSignatures
+    fromProto uis = do
+        signatures <-
+            fmap Map.fromAscList
+                . mapM deMk
+                . Map.toAscList
+                =<< uis ^? ProtoFields.signatures
+        return Updates.UpdateInstructionSignatures{..}
+      where
+        deMk (k, s) = do
+            sig <- fromProto s
+            return (fromIntegral k, sig)
+
+instance FromProto Proto.TransactionTime where
+    type Output' Proto.TransactionTime = TransactionTime
+    fromProto = return . TransactionTime . deMkWord64
+
+instance FromProto Proto.UpdateSequenceNumber where
+    type Output' Proto.UpdateSequenceNumber = Nonce
+    fromProto = return . Nonce . deMkWord64
+
+instance FromProto Proto.UpdateInstructionHeader where
+    type Output' Proto.UpdateInstructionHeader = Updates.UpdateHeader
+    fromProto uiHeader = do
+        updateSeqNumber <- fromProto =<< uiHeader ^? ProtoFields.sequenceNumber
+        updateEffectiveTime <- fromProto =<< uiHeader ^? ProtoFields.effectiveTime
+        updateTimeout <- fromProto =<< uiHeader ^? ProtoFields.timeout
+        let updatePayloadSize = PayloadSize 0 -- FIXME: What should go here, no such field in the protobuf equivalent.
+        return Updates.UpdateHeader{..}
+
+instance FromProto Proto.UpdateInstruction where
+    type Output' Proto.UpdateInstruction = Updates.UpdateInstruction
+    fromProto ui = do
+        uiSignHash <- Nothing -- FIXME: What should go here, no such field in the protobuf equivalent.
```

Comment:
`makeUpdateInstruction` has signature `RawUpdateInstruction -> Map.Map UpdateKeyIndex KeyPair -> UpdateInstruction`.

I'm not sure what to provide for this parameter. The payload has a field corresponding to `Map.Map UpdateKeyIndex Signature`. Should the `uiSignatures` of the `UpdateInstruction` be set manually to this?

Reply:
Right. Then we need to add a similar helper there that takes signatures instead of keypairs. Maybe it already exists in that module, cannot check now. The main point is that computing the hash should be done in the one place. Not here since it is tricky to make sure it is the same.